### PR TITLE
Debug/Fix CLI and OpenBCI Boards

### LINKED ISSUE #40: Add option to select OpenBCI usb dongle
It would be nice to select the cyton serial port, instead of having to type it each time.

Example com ports:
Mac - [DEFAULT]: OpenBCI Port Name = /dev/cu.usbserial-DM00D7TW
Windows - COM4
LINUX - /dev/ttyUSB0

This applies to Ganglion, Cyton, and CytonDaisy Boards.

### PR DESCRIPTION
**Pull request by retiutut**
- [x] Fixes #40
- [x] Add support for Ganglion+WiFi via CLI
- [x] Add newline characters as needed to make input prompts more readable and organized for users

@JohnGriffiths @JadinTredup @m9h

**Full Console Log using Input Prompt and Ganglion w/ BLED112 Dongle on Mac OS 10.15.7:**
```
(eeg-nb-main) Richard-Waltmans-iMac-4:eeg-notebooks Richard$ eegnb runexp -ip
run command line prompt script
Welcome to NeurotechX EEG Notebooks. 
Please enter the integer value corresponding to your EEG device: 
[0] None 
[1] Muse2016 
[2] Muse2 
[3] MuseS 
[4] OpenBCI Ganglion 
[5] OpenBCI Cyton 
[6] OpenBCI Cyton + Daisy 
[7] G.Tec Unicorn 
[8] BrainBit 
[9] Notion 1 
[10] Notion 2 
 [11] Synthetic 

Enter Board Selection: 4
Selected board OpenBCI Ganglion 

Please select your connection method:
[0] usb dongle 
[1] wifi shield 

Enter connection method: 0

Ganglion MAC Address (Press Enter to Autoscan): 

Please select which experiment you would like to run: 
[0] visual n170 
[1] visual p300 
[2] ssvep 

Enter Experiment Selection: 0
Selected experiment visual-N170 

Now, enter the duration of the recording (in seconds). 

Enter duration: 12

Next, enter the ID# of the subject you are recording data from. 

Enter subject ID#: 1

Next, enter the session number you are recording for. 

Enter session #: 1

Getting a list of available serial ports...
[0] /dev/cu.MyLovesAirPods-Wireless - n/a
[1] /dev/cu.UEBOOM2-LWACP - n/a
[2] /dev/cu.Bluetooth-Incoming-Port - n/a
[3] /dev/cu.usbserial-DM00D7TW - FT231X USB UART - FT231X USB UART
[4] /dev/cu.usbmodem11 - Low Energy Dongle - CDC data
Select Port(number): 4
[2020-12-02 23:20:09.628] [brainflow_logger] [info] incomming json: {
    "ip_address": "",
    "ip_port": 0,
    "ip_protocol": 0,
    "mac_address": "",
    "other_info": "",
    "serial_number": "",
    "serial_port": "/dev/cu.usbmodem11",
    "timeout": 0
}
[2020-12-02 23:20:09.629] [brainflow_logger] [info] use 15 as a timeout for device discovery and for callbacks
[2020-12-02 23:20:09.629] [brainflow_logger] [info] mac address is not specified, try to find ganglion without it

EEG device successfully connected!
Press [ENTER] when ready to begin...
21.4626         WARNING         User requested fullscreen with size [1600  900], but screen is actually [2560, 1440]. Using actual size
23.0569         WARNING         User requested fullscreen with size [1600  900], but screen is actually [2560, 1440]. Using actual size
[2020-12-02 23:20:20.491] [brainflow_logger] [info] use command b to start streaming
```

### CODE REVIEW

**Comment by retiutut** [comment]
Ganglion+WiFi on Mac OS 10.15.7:

```
(eeg-nb-main) Richard-Waltmans-iMac-4:eeg-notebooks Richard$ eegnb runexp -ip
run command line prompt script
Welcome to NeurotechX EEG Notebooks. 
Please enter the integer value corresponding to your EEG device: 
[0] None 
[1] Muse2016 
[2] Muse2 
[3] MuseS 
[4] OpenBCI Ganglion 
[5] OpenBCI Cyton 
[6] OpenBCI Cyton + Daisy 
[7] G.Tec Unicorn 
[8] BrainBit 
[9] Notion 1 
[10] Notion 2 
 [11] Synthetic 

Enter Board Selection: 4
Selected board OpenBCI Ganglion 

Please select your connection method:
[0] usb dongle 
[1] wifi shield 

Enter connection method: 1

Enter Ganglion+WiFi IP Address: 10.0.1.45

Please select which experiment you would like to run: 
[0] visual n170 
[1] visual p300 
[2] ssvep 

Enter Experiment Selection: 0
Selected experiment visual-N170 

Now, enter the duration of the recording (in seconds). 

Enter duration: 12

Next, enter the ID# of the subject you are recording data from. 

Enter subject ID#: 1

Next, enter the session number you are recording for. 

Enter session #: 1
[2020-12-03 01:00:02.078] [brainflow_logger] [info] incomming json: {
    "ip_address": "10.0.1.45",
    "ip_port": 6677,
    "ip_protocol": 0,
    "mac_address": "",
    "other_info": "",
    "serial_number": "",
    "serial_port": "",
    "timeout": 0
}
[2020-12-03 01:00:02.078] [brainflow_logger] [info] use 10 as http timeout
[2020-12-03 01:00:02.078] [brainflow_logger] [info] local ip addr is 10.0.1.59
[2020-12-03 01:00:04.115] [brainflow_logger] [info] configuration string {"delimiter":true,"ip":"10.0.1.59","latency":10000,"output":"raw","port":6677}
[2020-12-03 01:00:06.175] [brainflow_logger] [info] command string {"command":"~4"}
[2020-12-03 01:00:08.291] [brainflow_logger] [warning] If you change gain you may need to rescale EXG data

EEG device successfully connected!
Press [ENTER] when ready to begin...
28.7602         WARNING         User requested fullscreen with size [1600  900], but screen is actually [2560, 1440]. Using actual size
29.6554         WARNING         User requested fullscreen with size [1600  900], but screen is actually [2560, 1440]. Using actual size
(eeg-nb-main) Richard-Waltmans-iMac-4:eeg-notebooks Richard$ 
```

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,7 @@
 __pycache__
 
+eeg_notebooks.egg-info/
```
Comment (ErikBjare) [inline .gitignore]: This could be *.egg-info, but just nitpicking :)

```suggestion
*.egg-info/
```

File: eegnb/cli/introprompt.py [hunk 2]
```diff
@@ -40,57 +40,61 @@ def intro_prompt():
...
         print("Please select your connection method:\n"
               "[0] usb dongle \n"
               "[1] wifi shield \n")
-        connect_idx = input("Enter connection method:")
+        connect_idx = int(input("Enter connection method: "))
 
         # add "_wifi" suffix to the end of the board name for brainflow
         if connect_idx == 1:
             board_selection = board_selection + "_wifi"
+            ip_address = input("\nEnter Ganglion+WiFi IP Address: ")
```
Comment (ErikBjare) [inline eegnb/cli/introprompt.py:65]: Ganglion is not the only device that can be used with the WiFi shield.

```suggestion
            ip_address = input("\nEnter IP Address of device: ")
```

File: eegnb/cli/introprompt.py [hunk 3]
```diff
@@ -40,57 +40,61 @@ def intro_prompt():
...
+        else:
+            # if the ganglion is being used, you can enter optional Ganglion mac address
+            if board_selection == 'ganglion':
+                ganglion_mac_address = input("\nGanglion MAC Address (Press Enter to Autoscan): ")
```
Comment (ErikBjare) [inline eegnb/cli/introprompt.py]: I don't think this is specific to the ganglion.

```suggestion
            mac_address = input("\nEnter Bluetooth MAC Address (or empty to autoscan): ")
```

File: eegnb/cli/introprompt.py [hunk 4]
```diff
@@ -40,57 +40,61 @@ def intro_prompt():
...
 
     # start the EEG device
-    if board_selection == 'ganglion':
-        # if the ganglion is chosen a MAC address should also be proviced
-        eeg_device = EEG(device=board_selection, mac_addr=mac_address)
+    if board_selection.startswith('ganglion'):
+        if board_selection == 'ganglion_wifi':
+            eeg_device = EEG(device=board_selection, ip_addr=ip_address)
+        else: 
+            eeg_device = EEG(device=board_selection, mac_addr=ganglion_mac_address)
```
Comment (ErikBjare) [inline eegnb/cli/introprompt.py:97]: The ganglion really isn't special here (afaik). Looks like it's called the same anyway.

```suggestion
    if not board_selection.endswith('_wifi'):
        eeg_device = EEG(device=board_selection, mac_addr=mac_address)
```

File: eegnb/devices/utils.py [hunk 5]
```diff
@@ -75,3 +70,46 @@ def create_stim_array(timestamps, markers):
...
+
+    #result = []
+    #for port in ports:
+        #try:
+            #s = serial.Serial(port)
+            #s.close()
+            #result.append(port)
+        #except (OSError, serial.SerialException):
+            #pass
+    #return result
```
Comment (ErikBjare) [inline eegnb/devices/utils.py]:
```suggestion
```

**Review by ErikBjare — COMMENTED** [review]
Nice changes, some minor nitpicks/suggestions :)

File: eegnb/cli/introprompt.py (hunk 2 above)
Comment (retiutut) [reply eegnb/cli/introprompt.py:65]: This prompt did not exist before....

File: eegnb/cli/introprompt.py (hunk 3 above)
Comment (retiutut) [reply eegnb/cli/introprompt.py]: Only the text description for the prompt was changed.

File: eegnb/cli/introprompt.py (hunk 4 above)
Comment (retiutut) [reply eegnb/cli/introprompt.py:97]:
```
# start the EEG device
    if board_selection.startswith('ganglion'):
        if board_selection == 'ganglion_wifi':
            eeg_device = EEG(device=board_selection, ip_addr=ip_address)
        else: 
            eeg_device = EEG(device=board_selection, mac_addr=ganglion_mac_address)
    else:
        eeg_device = EEG(device=board_selection) <---Fixed this
```

File: eegnb/cli/introprompt.py (hunk 2 above)
Comment (JadinTredup) [reply eegnb/cli/introprompt.py:65]: Hey @ErikBjare although you technically can use the Cyton/daisy with the Wifi shield, it is not recommended by OpenBCI due to some issues with packet loss so we made the decision to only officially support it with the Ganglion in the library.

File: eegnb/cli/introprompt.py (hunk 3 above)
Comment (ErikBjare) [reply eegnb/cli/introprompt.py]: Yeah I noticed, just saw something that could be simplified and suggested it.

Comment (JadinTredup) [reply eegnb/cli/introprompt.py]: @ErikBjare the ganglion is the only device in brainflow which uses a MAC address.

File: eegnb/cli/introprompt.py (hunk 4 above)
Comment (JadinTredup) [reply eegnb/cli/introprompt.py:97]: See: https://github.com/NeuroTechX/eeg-notebooks/pull/41#discussion_r535452277

File: eegnb/cli/introprompt.py (hunk 3 above)
Comment (retiutut) [reply eegnb/cli/introprompt.py]: The most important change right here is that I have tested Ganglion+BLED112 without a mac address on my Mac. Works great even leaving this blank (thanks to BrainFlow)!

I expect this specific configuration to work just fine on all OS.

File: eegnb/cli/introprompt.py (hunk 2 above)
Comment (ErikBjare) [reply eegnb/cli/introprompt.py:65]: I knew about the OpenBCI warning, but didn't know about the decision to not even attempt to support it here. 

Makes sense, although wouldn't the flow as it looks right now let you choose Wi-Fi as a connection option for other OpenBCI devices?

File: .gitignore (hunk 1 above)
Comment (retiutut) [reply .gitignore]: Looked up about Python "eggs" and made this change!

File: eegnb/cli/introprompt.py (hunk 2 above)
Comment (retiutut) [reply eegnb/cli/introprompt.py:65]: @ErikBjare Good point. I was on the fence on whether or not to include Cyton+Wifi, but I can apply @JadinTredup's conclusion now. Editing now to enforce this.

Comment (ErikBjare) [reply eegnb/cli/introprompt.py:65]: Maybe add a code comment while you're at it :)

Comment (retiutut) [reply eegnb/cli/introprompt.py:65]: Here is the revised code block that handles all supported OpenBCI Boards via CLI w/ input prompts:

```
    # Handles connectivity selection if an OpenBCI board is being used
    if board_selection in ['cyton', 'cyton_daisy', 'ganglion']:

        # determine whether board is connected via Wifi or BLE
        print("Please select your connection method:\n"
              "[0] usb dongle \n"
              "[1] wifi shield \n")
        connect_idx = int(input("Enter connection method: "))

        # add "_wifi" suffix to the end of the board name for brainflow
        if connect_idx == 1:
            board_selection = board_selection + "_wifi"

        if board_selection == 'ganglion':
            # If the Ganglion is being used, you can enter optional Ganglion mac address
            ganglion_mac_address = input("\nGanglion MAC Address (Press Enter to Autoscan): ")    
        elif board_selection == 'ganglion_wifi':
            # IP address is required for this board configuration
            ip_address = input("\nEnter Ganglion+WiFi IP Address: ")
        elif board_selection == 'cyton_wifi' or board_selection == 'cyton_daisy_wifi':
            print(f"\n{boards[board_idx]} + WiFi is not supported. Please use the dongle that was shipped with the device.\n")
            exit()
```


Full Console Log:

```
(eeg-nb-main) Richard-Waltmans-iMac-4:eeg-notebooks Richard$ eegnb runexp -ip
run command line prompt script
Welcome to NeurotechX EEG Notebooks. 
Please enter the integer value corresponding to your EEG device: 
[0] None 
[1] Muse2016 
[2] Muse2 
[3] MuseS 
[4] OpenBCI Ganglion 
[5] OpenBCI Cyton 
[6] OpenBCI Cyton + Daisy 
[7] G.Tec Unicorn 
[8] BrainBit 
[9] Notion 1 
[10] Notion 2 
 [11] Synthetic 

Enter Board Selection: 6
Selected board OpenBCI Cyton + Daisy 

Please select your connection method:
[0] usb dongle 
[1] wifi shield 

Enter connection method: 1

OpenBCI Cyton + Daisy + WiFi is not supported. Please use the dongle that was shipped with the device.
```

**Review by ErikBjare — APPROVED** [review]

**Review by JadinTredup — APPROVED** [review]

**Pull request merged**